File: java/dec/State.java
/* Copyright 2015 Google Inc. All Rights Reserved.

   Distributed under MIT license.
   See file LICENSE for detail or copy at https://opensource.org/licenses/MIT
*/

package org.brotli.dec;

import static org.brotli.dec.RunningState.BLOCK_START;
import static org.brotli.dec.RunningState.CLOSED;
import static org.brotli.dec.RunningState.UNINITIALIZED;

import java.io.IOException;
import java.io.InputStream;

final class State {
  RunningState runningState = UNINITIALIZED;
  RunningState nextRunningState;
  final BitReader br = new BitReader();
  byte[] ringBuffer;
  final int[] blockTypeTrees = new int[3 * Huffman.HUFFMAN_MAX_TABLE_SIZE];
  final int[] blockLenTrees = new int[3 * Huffman.HUFFMAN_MAX_TABLE_SIZE];

  // Current meta-block header information.
  int metaBlockLength;
  boolean inputEnd;
  boolean isUncompressed;
  boolean isMetadata;

  final HuffmanTreeGroup hGroup0 = new HuffmanTreeGroup();
  final HuffmanTreeGroup hGroup1 = new HuffmanTreeGroup();
  final HuffmanTreeGroup hGroup2 = new HuffmanTreeGroup();
  final int[] blockLength = new int[3];
  final int[] numBlockTypes = new int[3];
  final int[] blockTypeRb = new int[6];
  final int[] distRb = {16, 15, 11, 4};
  int pos = 0;
  int maxDistance = 0;
  int distRbIdx = 0;
  boolean trivialLiteralContext = false;
  int literalTreeIndex = 0;
  int literalTree;
  int j;
  int insertLength;
  byte[] contextModes;
  byte[] contextMap;
  int contextMapSlice;
  int distContextMapSlice;
  int contextLookupOffset1;
  int contextLookupOffset2;
  int treeCommandOffset;
  int distanceCode;
  byte[] distContextMap;
  int numDirectDistanceCodes;
  int distancePostfixMask;
  int distancePostfixBits;
  int distance;
  int copyLength;
  int copyDst;
  int maxBackwardDistance;
  int maxRingBufferSize;
  int ringBufferSize = 0;
  long expectedTotalSize = 0;
  byte[] customDictionary = new byte[0];
  int bytesToIgnore = 0;

  int outputOffset;
  int outputLength;
  int outputUsed;
  int bytesWritten;
  int bytesToWrite;
  byte[] output;

  // TODO: Update to current spec.
  private static int decodeWindowBits(BitReader br) {
    if (BitReader.readBits(br, 1) == 0) {
      return 16;
    }
    int n = BitReader.readBits(br, 3);
    if (n != 0) {
      return 17 + n;
    }
    n = BitReader.readBits(br, 3);
    if (n != 0) {
      return 8 + n;
    }
    return 17;
  }

  /**
   * Associate input with decoder state.
   *
   * @param state uninitialized state without associated input
   * @param input compressed data source
   */
  static void setInput(State state, InputStream input) {
    if (state.runningState != UNINITIALIZED) {
      throw new IllegalStateException("State MUST be uninitialized");
    }
    BitReader.init(state.br, input);
    int windowBits = decodeWindowBits(state.br);
    if (windowBits == 9) { /* Reserved case for future expansion. */
      throw new BrotliRuntimeException("Invalid 'windowBits' code");
    }
    state.maxRingBufferSize = 1 << windowBits;
    state.maxBackwardDistance = state.maxRingBufferSize - 16;
    state.runningState = BLOCK_START;
  }

  static void close(State state) throws IOException {
    if (state.runningState == UNINITIALIZED) {
      throw new IllegalStateException("State MUST be initialized");
    }
    if (state.runningState == CLOSED) {
      return;
    }
    state.runningState = CLOSED;
    BitReader.close(state.br);
  }
}

File: csharp/CSharpBrotli/CSharpBrotli/Decode/State.cs
using System;
using System.IO;

namespace CSharpBrotli.Decode
{
    public sealed class State
    {
        public RunningStage runningState = RunningStage.UNINITIALIZED;
        public RunningStage nextRunningState;
        public readonly BitReader br = new BitReader();
        public byte[] ringBuffer;
        public readonly int[] blockTypeTrees = new int[3 * Huffman.HUFFMAN_MAX_TABLE_SIZE];
        public readonly int[] blockLenTrees = new int[3 * Huffman.HUFFMAN_MAX_TABLE_SIZE];

        // Current meta-block header information.
        public int metaBlockLength;
        public bool inputEnd;
        public bool isUncompressed;
        public bool isMetadata;

        public readonly HuffmanTreeGroup hGroup0 = new HuffmanTreeGroup();
        public readonly HuffmanTreeGroup hGroup1 = new HuffmanTreeGroup();
        public readonly HuffmanTreeGroup hGroup2 = new HuffmanTreeGroup();
        public readonly int[] blockLength = new int[3];
        public readonly int[] numBlockTypes = new int[3];
        public readonly int[] blockTypeRb = new int[6];
        public readonly int[] distRb = { 16, 15, 11, 4 };
        public int pos = 0;
        public int maxDistance = 0;
        public int distRbIdx = 0;
        public bool trivialLiteralContext = false;
        public int literalTreeIndex = 0;
        public int literalTree;
        public int j;
        public int insertLength;
        public byte[] contextModes;
        public byte[] contextMap;
        public int contextMapSlice;
        public int distContextMapSlice;
        public int contextLookupOffset1;
        public int contextLookupOffset2;
        public int treeCommandOffset;
        public int distanceCode;
        public byte[] distContextMap;
        public int numDirectDistanceCodes;
        public int distancePostfixMask;
        public int distancePostfixBits;
        public int distance;
        public int copyLength;
        public int copyDst;
        public int maxBackwardDistance;
        public int maxRingBufferSize;
        public int ringBufferSize = 0;
        public long expectedTotalSize = 0;
        public byte[] customDictionary = new byte[0];
        public int bytesToIgnore = 0;

        public int outputOffset;
        public int outputLength;
        public int outputUsed;
        public int bytesWritten;
        public int bytesToWrite;
        public byte[] output;

        private static int DecodeWindowBits(BitReader br)
        {
            if (BitReader.ReadBits(br, 1) == 0)
            {
                return 16;
            }
            int n = BitReader.ReadBits(br, 3);
            if (n != 0)
            {
                return 17 + n;
            }
            n = BitReader.ReadBits(br, 3);
            if (n != 0)
            {
                return 8 + n;
            }
            return 17;
        }

        public static void SetInput(State state, Stream input)
        {
            if (state.runningState != RunningStage.UNINITIALIZED)
            {
                throw new InvalidOperationException("State MUST be uninitialized");
            }
            BitReader.Init(state.br, input);
            int windowBits = DecodeWindowBits(state.br);
            if (windowBits == 9)
            { /* Reserved case for future expansion. */
                throw new BrotliRuntimeException("Invalid 'windowBits' code");
            }
            state.maxRingBufferSize = 1 << windowBits;
            state.maxBackwardDistance = state.maxRingBufferSize - 16;
            state.runningState = RunningStage.BLOCK_START;
        }

        public static void Close(State state)
        {
            try
            {
                if (state.runningState == RunningStage.UNINITIALIZED)
                {
                    throw new InvalidOperationException("State MUST be initialized");
                }
                if (state.runningState == RunningStage.CLOSED)
                {
                    return;
                }
                state.runningState = RunningStage.CLOSED;
                BitReader.Close(state.br);
            }
            catch(IOException ex)
            {
                throw ex;
            }
        }
    }
}